mm2src/mm2_main/src/lp_ordermatch.rs

```diff
@@ -1167,6 +1171,9 @@ pub struct TakerRequest {
     #[serde(default)]
     #[serde(skip_serializing_if = "Option::is_none")]
     pub rel_protocol_info: Option<Vec<u8>>,
+    #[serde(default)]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub swap_version: Option<u32>,
```

Comment:
I think this swap_version is needed to differentiate between legacy and TPU swaps.
I'd suggest making it a enum (Legacy/TPU) instead of numeric.
I believe a numeric 'version' is what we can change regularly (like improving protocol between nodes).

Reply:
> I'd suggest making it a enum (Legacy/TPU) instead of numeric.

TPU is already a numeric, we call ongoing update as Trading Protocol Upgrade V2 also we have swap v2 naming in repo.

> I believe a numeric 'version' is what we can change regularly (like improving protocol between nodes).

I think it doesn't matter is it regular or rare, every swap protocol update which breaks backward compatibility and forces all seed nodes to be updated requires the version bump.

I suggest to keep versioning simple, where we treat legacy as v1 or none.
I think I will make field non optional and provide necessary functions for `default` and `skip_serializing_if` annotations, to not do `None` and `Some(1)` checks for legacy

---
> where we treat legacy as v1

Also, its logically to think about legacy as v1 version as our tpu is starting from 2 not 1

Reply:
simplified swap_version field https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261/commits/707946a49fee99d7a074a0389bd4d4c10f155dbb

also please join the discussion if you think you have smth to add
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261#discussion_r1841572789

Reply:
I think legacy and TPU are not versions in fact, but two different algorithms.
(I am concerned, because I added a version field in another PR which will be increased when we add/remove features to the swap protocol. I think it may be confusing to have several version fields so adding a scope to it, where it possible, would help to better understand code)

Reply:
> (I am concerned, because I added a version field in another PR which will be increased when we add/remove features to the swap protocol. I think it may be confusing to have several version fields so adding a scope to it, where it possible, would help to better understand code)

Could you provide links with new versions fields please?

Also found these lines 
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2112/files#diff-21afdd8d37a9aa80dbf375468ad6f5f494436c419d27b5ff443425dfba5f10faR288-R292
![image](https://github.com/user-attachments/assets/2c279f56-dcfd-4c5a-bdab-7a5aff468241)

I think these constants already create confusion, as we call tpu as version 2 upgrade, while in utxo burn PR for upgraded version you use `1`.
I suggest to change legacy version to version 1 and use version 2 for new protocol.

---
Also I have this task for next sprint 
> add version in negotiation p2p msg and NegotiationData (to know what Maker/Taker version do we need in the recreate process).

I'm leaning to have another versioning field name for these types, not `swap_protocol`, as there could be situation, when we use same trading protocol version, just decided to add additional fields to messages (may be for protocol fixes or additional features).

Reply:
> > I think legacy and TPU are not versions in fact, but two different algorithms.
> 
> Ummm, we can consider them as backward incompatible versions with each other. Tbh i am all for it for having just one version field that controls minor features and swap protocol used, this is simpler than using multiple versions.

Are you planning to ask users to change the configuration every time we make minor changes to the protocol?
I think this should only be required for major changes, like protocol upgrades. Minor changes could be handled internally by KDF.

Reply:
> I think these constants already create confusion
That version in my PR is the protocol version in the negotiation message and is needed to determine if the remote node supports a new feature or not. For e.g. we add new burn tx output and remote node may not be upgraded so we don't use this feature for that node.
This protocol version could grow up within legacy swap protocol, so it's not for differentiating legacy and TPU, it's to differentiate old and new legacy swap nodes.
I suggest within TPU we should have a similar protocol_version in the negotiation message, this would allow to determine that the remote node is old (I think no need to start it from a specific number like 2 as it is defined within TPU and does not intersect with the same legacy protocol version)

Reply:
> That version in my PR is the protocol version in the negotiation message and is needed to determine if the remote node supports a new feature or not.

So its about messages versioning within some protocol, not protocol version itself? Correct me if I understood it wrong.

Then fields and constants names shouldn't have such naming like `swap_protocol_` or `legacy_protocol_`
```
pub const LEGACY_PROTOCOL_VERSION: u16 = 0;

pub const SWAP_PROTOCOL_VERSION: u16 = 1;
```

fields and constant names should reflect that its message version.

---
> This protocol version could grow up within legacy swap protocol, so it's not for differentiating legacy and TPU, it's to differentiate old and new legacy swap nodes.

I have same idea for p2p and negotiation messages in `taker_swap_v2` and `maker_swap_v2`. Messages could have different versions within the same swap protocol version

Reply:
Yes in my PR I should rename these:
```
pub const LEGACY_PROTOCOL_VERSION: u16 = 0;
pub const SWAP_PROTOCOL_VERSION: u16 = 1;
```
into
```
pub const LEGACY_SWAP_DEFAULT_PROTOCOL_VERSION: u16 = 0;
pub const LEGACY_SWAP_CURRENT_PROTOCOL_VERSION: u16 = 1;
```
I hope this will make it clearer

Reply:
> Yes in my PR I should rename these:
> 
> ```
> pub const LEGACY_PROTOCOL_VERSION: u16 = 0;
> pub const SWAP_PROTOCOL_VERSION: u16 = 1;
> ```
> 
> into
> 
> ```
> pub const LEGACY_SWAP_DEFAULT_PROTOCOL_VERSION: u16 = 0;
> pub const LEGACY_SWAP_CURRENT_PROTOCOL_VERSION: u16 = 1;
> ```
> 
> I hope this will make it clearer

just one note

may be this?
```
pub const LEGACY_SWAP_DEFAULT_MSG_VERSION: u16 = 0;
pub const LEGACY_SWAP_CURRENT_MSG_VERSION: u16 = 1;
```

or what protocol are you providing inside legacy swap protocol?

Reply:
> So its about messages versioning within some protocol, not protocol version itself? Correct me if I understood it wrong.

well, in daemons code they call it 'protocol version' and change each time whenever any, even slight-less, change in p2p messages occurs, this number just grows steadily.
If we want it to reflect that this is legacy or tpu we need to make it structured to provide space for numbering change (but this is may be not needed as both message flows do not mix)

Reply:
> That might be in the swap negotiation layer or better in the order matching layer

I implemented this for legacy at the negotiation level in #2112
I guess for order matching we would need this as well (if we may add new swap types or features)

Reply:
> I guess for order matching we would need this as well (if we may add new swap types or features)
But I think for order matching versoning will change more rarely

I meant it would be better off having such negotiation params agreed upon before matching the order to the taker. This doesn't really shine in the burn case since we consider it as optional feature (if the other peer doesn't support it we will just proceed without burn), but if we wanna mark the feature/fix as required, then such a swap will fail after starting, in which case not starting it was the efficient choice to avoid locking funds etc...

Reply:
As for ordermatching, this pr introduces `swap_version` in Taker and Maker orders

Reply:
Resolving as swap_version was removed.
(I guess we are still debating on its implementation. Personally I am sure the version field is a must in any communication protocol, even if we think we do not use it for now)